storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3088,6 +3119,690 @@ std::ostream &fseg_header::to_stream(std::ostream &out) const
+  allocate the new extent for the given segment.
+  If the extent descriptor state is XDES_FSEG then
+  try to find the free page from the segment.
+  @param seg_header segment header where new extent will be added
+  @param iblock     inode block
+  @param inode      index node information
+  @param extent     create a new extent
+  @param mtr        mini-transaction
+  @param err        error code
+  @return block or nullptr in case of failure */
+  buf_block_t *alloc_page_from_new_seg(
+    fseg_header_t *seg_header, buf_block_t *iblock,
+    fseg_inode_t *inode, uint32_t extent, mtr_t *mtr,
+    dberr_t *err)
+  {
+    fil_space_t *space= fil_system.sys_space;
+    buf_block_t *fsp_header= fsp_get_header(space, mtr, err);
+    if (*err)
+      return nullptr;
+    uint64_t seg_id= mach_read_from_8(inode + FSEG_ID);
+    buf_block_t *xdes;
+    xdes_t *descr=
+      xdes_get_descriptor_with_space_hdr(fsp_header, space, extent, mtr,
+                                         err, &xdes);
+    uint32_t allocated_page= FIL_NULL;
+
+    if (!descr)
+      return nullptr;
+
+    ut_a(xdes_get_state(descr) == XDES_FSEG
+         || xdes_get_state(descr) == XDES_FREE);
+
+    /* If the descriptor already belongs to the segment
+    then allocate the free page */
+    if (xdes_get_state(descr) == XDES_FSEG
+        && mach_read_from_8(descr + XDES_ID) == seg_id)
+    {
+      allocated_page= xdes_find_free(descr);
+      if (allocated_page == FIL_NULL)
+      {
+        *err= DB_ERROR;
+	return nullptr;
+      }
+      allocated_page += xdes_get_offset(descr);
+    }
+    else
+    {
+      descr= fsp_alloc_free_extent(space, extent, &xdes, mtr, err);
+      xdes_set_state(*xdes, descr, XDES_FSEG, mtr);
+      mtr->write<8,mtr_t::MAYBE_NOP>(*xdes, descr + XDES_ID, seg_id);
+      *err= flst_add_last(
+              iblock,
+              static_cast<uint16_t>(inode - iblock->page.frame
+                                    + FSEG_FREE), xdes,
+              static_cast<uint16_t>(descr - xdes->page.frame
+                                    + XDES_FLST_NODE),
+              space->free_limit, mtr);
+      if (*err)
+        return nullptr;
+
+      allocated_page= extent;
+      *err = fseg_fill_free_list(inode, iblock, space,
+                                 extent + FSP_EXTENT_SIZE, mtr);
+    }
+
+    if (*err == DB_SUCCESS)
+    {
+      *err= fseg_mark_page_used(space, inode, iblock, allocated_page,
+                                descr, xdes, mtr);
+      if (*err == DB_SUCCESS)
+        return fsp_page_create(space, allocated_page, mtr);
+    }
+    return nullptr;
+  }
+};
+
+void indexDefragmenter::store_child_pages(buf_block_t *block)
+{
+  mem_heap_t *heap= nullptr;
+  rec_offs *offsets= nullptr;
+  rec_t *rec=
+    page_rec_get_next(page_get_infimum_rec(buf_block_get_frame(block)));
+  while (!page_rec_is_supremum(rec))
+  {
+    offsets= rec_get_offsets(rec, m_index, offsets, 0,
+                             ULINT_UNDEFINED, &heap);
```

Comment:
The system tables should always use `ROW_FORMAT=REDUNDANT`. Therefore, we can use simpler access methods here, and there is no need to invoke `rec_get_offsets()`.